lib/elixir/test/elixir/kernel/warning_test.exs

```diff
@@ -452,6 +452,30 @@ defmodule Kernel.WarningTest do
     purge(RedefineSample)
   end
 
+  test "unused variable because re-declared in a match? pattern" do
+    assert_warn_eval(
+      [
+        "nofile:1:16",
+        "variable \"x\" is unused (there is a variable with the same name in the context,",
+        "variable \"x\" is unused (if the variable is not meant to be used,"
+      ],
+      """
+      fn x -> match?(x, :value) end
+      """
+    )
+
+    assert_warn_eval(
+      [
+        "nofile:1",
+        "variable \"&1\" is unused (there is a variable with the same name in the context,",
+        "variable \"&1\" is unused (if the variable is not meant to be used,"
```

Comment:
Note: there might be smarter warnings to emit in this case, but for now the test documents the existing behavior and I didn't went down this rabbit hole.

Reply:
There is an issue here is that we suggest the variable to be renamed with underscore, which is not possible. So we may want to check if the variable starts with `&` and skip that portion of the warning?